src/mlpack/methods/ann/augmented/ham_unit_impl.hpp

```diff
@@ -0,0 +1,90 @@
+/**
+ * @file ham_unit_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of HAMUnit class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_AUGMENTED_HAM_UNIT_IMPL_HPP
+#define MLPACK_METHODS_ANN_AUGMENTED_HAM_UNIT_IMPL_HPP
+
+#include "ham_unit.hpp"
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+
+template<typename E, typename J, typename S, typename W>
+HAMUnit<E, J, S, W>::HAMUnit(size_t memorySize,
+                 size_t memoryDim,
+                 E& embed,
+                 J& join,
+                 S& search,
+                 W& write)
+  : memorySize(memorySize), memoryDim(memoryDim),
+    search(search), embed(embed), t(0)
+{
+  memory = TreeMemory<double, J, S>(memorySize, join, write);
```

Comment:
Since there is no default constructor for  ``TreeMemory`` we should use the initialization list:

```c++
...
embed(embed),
t(0), 
memory(TreeMemory<double, J, S>(memorySize, memoryDim, join, write))
```

Also you missed the ``memoryDim`` parameter.